qiskit/circuit/quantumcircuit.py

```diff
@@ -4683,6 +4685,202 @@ def mcp(
+            _apply_cu(
+                self,
+                theta,
+                -pi / 2,
+                pi / 2,
+                control_qubits[0],
+                target_qubit,
+                use_basis_gates=use_basis_gates,
+            )
+        elif n_c < 4:
+            theta_step = theta * (1 / (2 ** (n_c - 1)))
+            _apply_mcu_graycode(
+                self,
+                theta_step,
+                -pi / 2,
+                pi / 2,
+                control_qubits,
+                target_qubit,
+                use_basis_gates=use_basis_gates,
+            )
+        else:
+            cgate = _mcsu2_real_diagonal(
+                RXGate(theta),
+                num_controls=len(control_qubits),
+                use_basis_gates=use_basis_gates,
+            )
+            self.compose(cgate, control_qubits + [target_qubit], inplace=True)
+
+    def mcry(
+        self,
+        theta: ParameterValueType,
+        q_controls: Union[QuantumRegister, List[Qubit]],
+        q_target: Qubit,
+        q_ancillae: Optional[Union[QuantumRegister, Tuple[QuantumRegister, int]]] = None,
+        mode: Optional[str] = None,
+        use_basis_gates: bool = False,
+    ):
+        """
+        Apply Multiple-Controlled Y rotation gate
+
+        Args:
+            self (QuantumCircuit): The QuantumCircuit object to apply the mcry gate on.
+            theta (float): angle theta
+            q_controls (list(Qubit)): The list of control qubits
+            q_target (Qubit): The target qubit
+            q_ancillae (QuantumRegister or tuple(QuantumRegister, int)): The list of ancillary qubits.
+            mode (string): The implementation mode to use
+            use_basis_gates (bool): use p, u, cx
+
+        """
+        # pylint: disable=cyclic-import
+        from .library.standard_gates.ry import RYGate
+        from .library.standard_gates.x import MCXGate
+        from qiskit.synthesis.multi_controlled import (
+            _apply_cu,
+            _apply_mcu_graycode,
+            _mcsu2_real_diagonal,
+        )
+
+        control_qubits = self._qbit_argument_conversion(q_controls)
+        target_qubit = self._qbit_argument_conversion(q_target)
+        if len(target_qubit) != 1:
+            raise QiskitError("The mcry gate needs a single qubit as target.")
+        ancillary_qubits = [] if q_ancillae is None else self._qbit_argument_conversion(q_ancillae)
+        all_qubits = control_qubits + target_qubit + ancillary_qubits
+        target_qubit = target_qubit[0]
+        self._check_dups(all_qubits)
+
+        # auto-select the best mode
+        if mode is None:
+            # if enough ancillary qubits are provided, use the 'v-chain' method
+            additional_vchain = MCXGate.get_num_ancilla_qubits(len(control_qubits), "v-chain")
+            if len(ancillary_qubits) >= additional_vchain:
+                mode = "basic"
+            else:
+                mode = "noancilla"
+
+        if mode == "basic":
```

Comment:
This is not something to fix in this PR: but do you know why we have this special case only for `MCRY` and not for the others? If it's more efficient we could also use it there. 

This is something we can fix very cleanly if we synthesize controlled gates via HLS, which knows the number of free auxiliaries.

Reply:
The use of ancillas was done in a PR that you implemented https://github.com/Qiskit/qiskit/pull/4787 following issue https://github.com/Qiskit/qiskit/issues/4786. I'm not sure if there is some improvement anymore.

Reply:
after checking it seems that adding ancillas can improve mcry in certain cases (num_controls >= 7). This is an optimization that should be explored in a follow-up work.